### DISCUSSION TITLE
Making multiple targets concurrently: is it possible?

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#865] [Help] [answered]
I am using Rstudio server/workbench but some process take a long time and I don't want to stop working on Rstudio so I run some targets via the terminal in the server.

The command is like this: 

`nohup /opt/R/4.1.2/bin/R -e "targets::tar_make(c('target1', 'target2'), callr_function = NULL)" > run_make.Rout 2>&1 &`

I continued working on RStudio, and I started a new target on the Rstudio terminal `targets::tar_make(target3)` that did not depend on target1, target2 and all the previous targets were up to date.
Then it got me thinking and I stopped that process.

Now I see that `project/_targets/meta/process` shows the id a of the old canceled process for target3. 

```
name|value
pid|2678092
version_r|4.1.2
version_targets|0.11.0
```

Is there a way to run multiple targets at the same time? 
Will the initial jobs, that are still running, be recognized as up to date by targets? Currently `tar_visnetwork` only shows the canceled task of targets3 as started but not the other targets 1 and 2. 

Note: this is a single machine with multiple cores and I've done this to speed up some process while not fully using the server and I could wait till targets 1 and 2 are finished to make target 3.

**Comment by wlandau** [comment, marked as answer]
Yes, you can definitely run conditionally independent targets in parallel. It is a core feature of the package, and it is explained at https://books.ropensci.org/targets/hpc.html. You will add a little more config to _targets.R and run tar_make_clustermq() or tar_make_future(). Both can utilize the cores of a local machine.

**Reply by llrs** [reply]
Many thanks for the quick reply! Yes, I know that I can use parallel targets and as far as I understood that chapter using clustermq or future configuration will always run (all) targets in parallel. But I don't need that as most of the steps are fast enough and the function to build target1 and target2 is already parallelized internally (but configurable via a parameter) so I hope this configuration won't cause troubles. However, if I have a tar_make running can I start a new tar_make concurrently? Or should I avoid this kind of workflow?

What about the process that is still running but not detected by tar_visnetwork? Will the output be recognized by targets or do I need to cancel the job and start it again?

**Reply by wlandau** [reply]
You can set deployment =”main” in tar_target() for any targets you don’t want to send to parallel workers. The package is not designed for concurrent instances of tar_make(), and it can cause problems.

> Will the initial jobs, that are still running, be recognized as up to date by targets? Currently tar_visnetwork only shows the canceled task of targets3 as started but not the other targets 1 and 2.

I am not sure what you mean, but it sounds like a problem that could happen if you have concurrent tar_make()s that are competing to modify the same data store.

I am out of office until July 20 and can help troubleshoot when I get back.

**Reply by llrs** [reply]
Don't worry, it is the concurrent problem and I stopped all jobs and restarted just one, I made sure it is detected by `tar_visnetwork()`. Thanks for the tip on `deployment` I should read the documentation more carefully. Enjoy your time off!
